src/Open3D/Visualization/Rendering/Filament/FilamentRenderer.cpp

```diff
@@ -0,0 +1,272 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+// IN THE SOFTWARE.
+// ----------------------------------------------------------------------------
+
+#include "FilamentRenderer.h"
+
+#include "Open3D/Utility/Console.h"
+
+#include <filament/Engine.h>
+#include <filament/LightManager.h>
+#include <filament/RenderableManager.h>
+#include <filament/Renderer.h>
+#include <filament/Scene.h>
+#include <filament/SwapChain.h>
+
+#include "FilamentCamera.h"
+#include "FilamentEntitiesMods.h"
+#include "FilamentRenderToBuffer.h"
+#include "FilamentResourceManager.h"
+#include "FilamentScene.h"
+#include "FilamentView.h"
+
+namespace open3d {
+namespace visualization {
+
+FilamentRenderer::FilamentRenderer(filament::Engine& aEngine,
+                                   void* nativeDrawable,
+                                   FilamentResourceManager& aResourceManager)
+    : engine_(aEngine), resourceManager_(aResourceManager) {
+    swapChain_ = engine_.createSwapChain(nativeDrawable);
+    renderer_ = engine_.createRenderer();
+
+    materialsModifier_ = std::make_unique<FilamentMaterialModifier>();
+}
+
+FilamentRenderer::~FilamentRenderer() {
+    scenes_.clear();
+
+    engine_.destroy(renderer_);
+    engine_.destroy(swapChain_);
+}
+
+SceneHandle FilamentRenderer::CreateScene() {
+    auto handle = SceneHandle::Next();
+    scenes_[handle] =
+            std::make_unique<FilamentScene>(engine_, resourceManager_);
+
+    return handle;
+}
+
+Scene* FilamentRenderer::GetScene(const SceneHandle& id) const {
+    auto found = scenes_.find(id);
+    if (found != scenes_.end()) {
+        return found->second.get();
+    }
+
+    return nullptr;
+}
+
+void FilamentRenderer::DestroyScene(const SceneHandle& id) {
+    scenes_.erase(id);
+}
+
+void FilamentRenderer::UpdateSwapChain() {
+    void* nativeWindow = swapChain_->getNativeWindow();
+    engine_.destroy(swapChain_);
+
+#if defined(__APPLE__)
+    auto resizeMetalLayer = [](void* nativeWindow) -> void* {
+        utility::LogError(
+                "::resizeMetalLayer() needs to be implemented. Please see "
+                "filament/samples/app/NativeWindowHelperCocoa.mm for "
+                "reference.");
+        return nativeWindow;
+    };
+
+    void* nativeSwapChain = nativeWindow;
+    void* metalLayer = nullptr;
+    auto backend = engine_.getBackend();
+    if (backend == filament::Engine::Backend::METAL) {
+        metalLayer = resizeMetalLayer(nativeWindow);
+        // The swap chain on Metal is a CAMetalLayer.
+        nativeSwapChain = metalLayer;
+    }
+
```

Comment:
I don't understand. If the backend is METAL we have code so that we do NOT break. (In theory). Likewise for Vulkan. But you have to manually change the code to use those backends. The other backend is OPENGL, which obviously we do support.